internal/provider/general_api_models.go

```diff
@@ -16,3 +31,227 @@ type SummaryAPIModel struct {
+	Id               types.Int64                      `tfsdk:"id"`
+	Name             types.String                     `tfsdk:"name"`
+	Organization     types.Int64                      `tfsdk:"organization"`
+	OrganizationName types.String                     `tfsdk:"organization_name"`
+	Url              types.String                     `tfsdk:"url"`
+	NamedUrl         types.String                     `tfsdk:"named_url"`
+	Description      types.String                     `tfsdk:"description"`
+	Variables        customtypes.AAPCustomStringValue `tfsdk:"variables"`
+}
+
+// This function allows us to parse the incoming data in HTTP requests from the API
+// into the BaseDataSourceModel instances.
+func (d *BaseDataSourceModel) ParseHttpResponse(body []byte) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	// Unmarshal the JSON response
+	var apiWorkflowJobTemplate WorkflowJobTemplateAPIModel
+	err := json.Unmarshal(body, &apiWorkflowJobTemplate)
+	if err != nil {
+		diags.AddError("Error parsing JSON response from AAP", err.Error())
+		return diags
+	}
+
+	// Map response to the WorkflowJobTemplate datesource schema
+	d.Id = types.Int64Value(apiWorkflowJobTemplate.Id)
+	d.Organization = types.Int64Value(apiWorkflowJobTemplate.Organization)
+	d.OrganizationName = ParseStringValue(apiWorkflowJobTemplate.SummaryFields.Organization.Name)
+	d.Url = ParseStringValue(apiWorkflowJobTemplate.Url)
+	d.NamedUrl = ParseStringValue(apiWorkflowJobTemplate.Related.NamedUrl)
+	d.Name = ParseStringValue(apiWorkflowJobTemplate.Name)
+	d.Description = ParseStringValue(apiWorkflowJobTemplate.Description)
+	d.Variables = ParseAAPCustomStringValue(apiWorkflowJobTemplate.Variables)
+
+	return diags
+}
+
+type StringDescriptions struct {
+	ApiEntitySlug         string
+	DescriptiveEntityName string
+	MetadataEntitySlug    string
+}
+
+// A struct to represent a base DataSource object, with a client and the slug name of
+// the API entity.
+type BaseDataSource struct {
+	client ProviderHTTPClient
+	StringDescriptions
+}
+
+// Constructs a new BaseDataSource object provided with a client instance (usually
+// initialized to nil, it will be later configured calling the Configure function)
+// and an apiEntitySlug string indicating the entity path name to consult the API.
+func NewBaseDataSource(client ProviderHTTPClient, stringDescriptions StringDescriptions) *BaseDataSource {
+	return &BaseDataSource{
+		client:             client,
+		StringDescriptions: stringDescriptions,
+	}
+}
+
+// Read refreshes the Terraform state with the latest data.
+func (d *BaseDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var state BaseDataSourceModel
+	var diags diag.Diagnostics
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	uri := path.Join(d.client.getApiEndpoint(), d.ApiEntitySlug)
```

Comment:
The constructor states that the `client`s default is nil, 

> client instance (usually initialized to nil, it will be later configured calling the Configure function)

Lets check to ensure the client exists before trying to call a function as we might get a problem here.